File: L1/mocks/MarketFeed.cpp
#include <iostream>
#include <chrono>
#include <vector>
#include <cstring>
#include <sys/socket.h>
#include <unistd.h>

struct MarketData {
    uint64_t timestamp;
    double price;
    uint32_t volume;
};

// Parsing function (works on raw bytes, no extra allocations)
inline MarketData parse(const char* buffer) {
    MarketData data;
    std::memcpy(&data, buffer, sizeof(MarketData));
    return data;
}

int main() {
    char buffer[sizeof(MarketData)];
    int sock = socket(AF_INET, SOCK_STREAM, 0);
    // Assume already connected...

    auto start = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < 1000000; i++) {
        read(sock, buffer, sizeof(buffer));
        MarketData md = parse(buffer);
        // Decision logic here (fast math, no heap allocation)
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::cout << "Elapsed: "
              << std::chrono::duration_cast<std::chrono::microseconds>(end - start).count()
              << " us\n";

    close(sock);
    return 0;
}

File: L1/mocks/MarketFeed.java
import java.io.*;
import java.net.Socket;

class MarketData {
    long timestamp;
    double price;
    int volume;
}

public class MarketFeed {
    public static void main(String[] args) throws Exception {
        Socket socket = new Socket("localhost", 5555);
        DataInputStream dis = new DataInputStream(socket.getInputStream());
        byte[] buffer = new byte[20];

        long start = System.nanoTime();

        for (int i = 0; i < 1_000_000; i++) {
            dis.readFully(buffer);
            MarketData md = parse(buffer); // creates new object each time
            // Decision logic
        }

        long end = System.nanoTime();
        System.out.println("Elapsed: " + (end - start)/1000 + " us");
        socket.close();
    }

    private static MarketData parse(byte[] buffer) throws IOException {
        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(buffer));
        MarketData md = new MarketData();
        md.timestamp = dis.readLong();
        md.price = dis.readDouble();
        md.volume = dis.readInt();
        return md;
    }
}

